modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/task/QueryTasksQueue.java

```diff
@@ -0,0 +1,376 @@
+     */
+    private static class Node {
+        /** */
+        QueryAwareTask item;
+
+        /** Next node in chain. */
+        Node next;
+
+        /** */
+        Node(QueryAwareTask item) {
+            this.item = item;
+        }
+    }
+
+    /** Current number of elements */
+    private final AtomicInteger cnt = new AtomicInteger();
+
+    /** Head of linked list. */
+    Node head;
+
+    /** Tail of linked list. */
+    private Node last;
+
+    /** */
+    private final ReentrantLock lock = new ReentrantLock();
+
+    /** Wait condition for waiting takes. */
+    private final Condition notEmpty = lock.newCondition();
+
+    /** Set of blocked (currently running) queries. */
+    private final Set<QueryKey> blockedQrys = new HashSet<>();
+
+    /**
+     * Creates a {@code LinkedBlockingQueue}.
+     */
+    public QueryTasksQueue() {
+        last = head = new Node(null);
+    }
+
+    /** Queue size. */
+    public int size() {
+        return cnt.get();
+    }
+
+    /** Add a task to the queue. */
+    public void addTask(QueryAwareTask task) {
+        lock.lock();
+
+        try {
+            enqueue(new Node(task));
+
+            cnt.getAndIncrement();
+
+            notEmpty.signal();
+        }
+        finally {
+            lock.unlock();
+        }
+    }
+
+    /** Poll task and block query. */
+    public QueryAwareTask pollTaskAndBlockQuery(long timeout, TimeUnit unit) throws InterruptedException {
+        lock.lockInterruptibly();
+
+        try {
+            QueryAwareTask res;
+
+            long nanos = unit.toNanos(timeout);
+            final AtomicInteger cnt = this.cnt;
+
+            while (cnt.get() == 0 || (res = dequeue()) == null) {
+                if (nanos <= 0L)
+                    return null;
+
+                nanos = notEmpty.awaitNanos(nanos);
+            }
+
+            blockedQrys.add(res.queryKey());
+
+            if (cnt.decrementAndGet() > 0)
+                notEmpty.signal();
+
+            return res;
+        }
+        finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Links node at end of queue.
+     *
+     * @param node The node.
+     */
+    private void enqueue(Node node) {
+        assert lock.isHeldByCurrentThread();
+        assert last.next == null : "Unexpected last.next: " + last.next;
+
+        last = last.next = node;
+    }
+
+    /**
+     * Removes a first non-blocked task from the head of the queue.
+     *
+     * @return The task.
+     */
+    private QueryAwareTask dequeue() {
+        assert lock.isHeldByCurrentThread();
+        assert head.item == null : "Unexpected head.item: " + head.item;
+
+        for (Node pred = head, cur = pred.next; cur != null; pred = cur, cur = cur.next) {
+            if (!blockedQrys.contains(cur.item.queryKey())) { // Skip tasks for blocked queries.
+                QueryAwareTask res = cur.item;
+
+                unlink(cur, pred);
+
+                return res;
+            }
+        }
```

Comment:
Let's rewite this with the while as it simplify the flow.

```suggestion
        Node pred = head;
        Node cur = pred.next;

        while (cur != null) {
            if (!blockedQrys.contains(cur.item.queryKey())) { // Skip tasks for blocked queries.
                QueryAwareTask res = cur.item;

                unlink(cur, pred);

                return res;
            }

            pred = cur;
            cur = cur.next;
        }
```

Reply:
It's a matter of taste, I think there is no simplification with `while`